src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c

```diff
@@ -21,35 +47,378 @@ const ActorInit En_Kakasi3_InitVars = {
+            this->unk_1AE *= -1.0f;
+        }
+
+        if (this->unk_1A4 != 0) {
+            this->actor.shape.rot.y += 0x1000;
+            if (this->actor.shape.rot.y == 0) {
+                this->unk_1A4 = 0;
+            }
+        }
+        currentFrame = this->skelAnime.animCurrentFrame;
+        if (currentFrame == 11 || currentFrame == 17) {
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_KAKASHI_SWING);
+        }
+        SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+    }
+}
+
+void func_80A911F0(EnKakasi3* this, GlobalContext* globalCtx) {
+    f32 frameCount = SkelAnime_GetFrameCount(&D_06000214.genericHeader);
+
+    SkelAnime_ChangeAnim(&this->skelAnime, &D_06000214, 1.0f, 0.0f, (s16)frameCount, 0, -10.0f);
+    this->actionFunc = func_80A91284;
+}
+
+void func_80A91284(EnKakasi3* this, GlobalContext* globalCtx) {
+    SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+
+    this->actor.textId = 0x40A1;
+    this->dialogState = 6;
+    this->unk_19A = 0;
+
+    if (LINK_IS_CHILD) {
+        this->unk_194 = 0;
+        if (gSaveContext.unk_12C5) {
+            this->actor.textId = 0x40A0;
+            this->dialogState = 5;
+            this->unk_1A8 = 1;
+        }
+    } else {
+        this->unk_194 = 1;
+        if (gSaveContext.unk_12C5) {
+            if (this->unk_195) {
+                this->actor.textId = 0x40A2;
+            } else {
+                this->actor.textId = 0x40A3;
+            }
+        }
+    }
+    this->actionFunc = func_80A91348;
+}
+
+void func_80A91348(EnKakasi3* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    s16 angleTowardsLink;
+    s16 absAngleTowardsLink;
+
+    func_80A90E28(this);
+    SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+    this->camId = -1;
+    if (func_8002F194(&this->actor, globalCtx)) {
+        if (this->unk_194 == 0) {
+            if (this->unk_1A8 == 0) {
+                this->actionFunc = func_80A91284;
+            } else {
+                this->actionFunc = func_80A91760;
+            }
+        } else {
+            this->actionFunc = func_80A91284;
+        }
+        return;
+    }
+
+    angleTowardsLink = this->actor.yawTowardsLink - this->actor.shape.rot.y;
+    if (!(this->actor.xzDistFromLink > 120.0f)) {
+        absAngleTowardsLink = ABS(angleTowardsLink);
+
+        if (absAngleTowardsLink < 0x4300) {
+            if (this->unk_194 == 0) {
+
+                if (player->stateFlags2 & 0x1000000) {
+                    this->camId = func_800800F8(globalCtx, 0x8D4, -0x63, &this->actor, 0);
+                    globalCtx->msgCtx.msgMode = 0x37;
+                    this->dialogState = 5;
+                    this->unk_1B8 = 0.0f;
+                    func_8010B680(globalCtx, 0x40A4, NULL);
+                    player->stateFlags2 |= 0x800000;
+                    this->actionFunc = func_80A915B8;
+                    return;
+                } else if (this->actor.xzDistFromLink < 80.0f) {
```

Comment:
No need for the else if there is already an early return imo

Reply:
These two early returns are necessary, otherwise the flow is incorrect, notice the call to func_8002F2CC down below.

Reply:
Right, but I'm not saying the early returns should go away, only the `else`s

Reply:
Oh, I misunderstood, you're right